exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp

```diff
@@ -0,0 +1,219 @@
+
+        kv_gains_[i] = _inv * Bk.transpose();
+        k_gains_[i] = _inv * Bk.transpose() * Sk * Ak;
+        ku_gains_[i] = _inv * R;
+        Sk = Ak.transpose() * Sk * (Ak - Bk * k_gains_[i]) + Q;
+
+        vk_gains_[i] = ((Ak - Bk * k_gains_[i]).transpose() * vk_gains_[i + 1]) -
+                     (k_gains_[i].transpose() * R * u) + (Q * x);
+
+        // fix for large values
+        Sk = Sk.unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
+            return std::min(std::max(x, k_min_clamp_), k_max_clamp_);
+        });
+        vk_gains_[i] = vk_gains_[i].unaryExpr([k_min_clamp_, k_max_clamp_](double x) -> double {
+            return std::min(std::max(x, k_min_clamp_), k_max_clamp_);
+        });
+    }
+}
+
+double ILQRSolver::ForwardPass(double alpha, Eigen::MatrixXdRef ref_trajectory)
+{
+    double cost = 0;
+    DynamicsSolverPtr dynamics_solver = prob_->GetScene()->GetDynamicsSolver();
+    for (int i = 0; i < T - 1; ++i)
+    {
+        Eigen::MatrixXd u = prob_->get_U(i);
+        // eq. 12
+        Eigen::MatrixXd delta_uk = -ku_gains_[i] * u - kv_gains_[i] * vk_gains_[i + 1] - k_gains_[i] * dynamics_solver->StateDelta(prob_->get_X(i), ref_trajectory.col(i));
+
+        u = u + alpha * delta_uk;
+        // clamp controls
+        if (parameters_.ControlLimits.size() > 0)
+            u = u.cwiseMax(-parameters_.ControlLimits).cwiseMin(parameters_.ControlLimits);
+
+        prob_->Update(u, i);
+        cost += prob_->GetControlCost(i) + prob_->GetStateCost(i);
+    }
+
+    // add terminal cost
+    cost += prob_->GetStateCost(T - 1);
+    return cost;
+}
+
+void ILQRSolver::Solve(Eigen::MatrixXd& solution)
+{
+    prob_->ResetCostEvolution(GetNumberOfMaxIterations() + 1);
+    if (!prob_) ThrowNamed("Solver has not been initialized!");
+
+    // all of the below are not pointers, since we want to copy over
+    //  solutions across iterations
+    Eigen::MatrixXd best_U, best_X, global_best_U;
+    solution.resize(ku_gains_[0].rows(), T);
+
+    HIGHLIGHT_NAMED("ILQRSolver", "Running ILQR solver for max " << parameters_.MaxIterations << " iterations");
+
+    double last_cost = 0, global_best_cost = 0;
+    int last_best_iteration = 0;
+
+    for (int i = 0; i < T - 1; ++i)
+        prob_->Update(Eigen::MatrixXd::Zero(prob_->get_U(0).rows(), prob_->get_U(0).cols()), i);
+
+    for (int iteration = 0; iteration < GetNumberOfMaxIterations(); ++iteration)
+    {
+        // Backwards pass computes the gains
+        BackwardPass();
+        HIGHLIGHT_NAMED("ILQRSolver", "Backward pass complete");
+
+        // forward pass to compute new control trajectory
+        auto alpha_space = Eigen::VectorXd::LinSpaced(10, 0.1, 1.0);
+        double best_cost = 0, best_alpha = 0;
+
+        Eigen::MatrixXd ref_trajectory = prob_->get_X();
+
+        // perform a linear search to find the best rate
+        for (int ai = 0; ai < alpha_space.rows(); ++ai)
```

Comment:
What you are doing is a forward line-search when we may try a backtracking line search for a performance improvement later - this just as an aside.